sycl/include/sycl/ext/intel/experimental/pipes.hpp

```diff
@@ -0,0 +1,213 @@
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if (_control_type == type::none) {
+      _type = 0;
+    } else if (_control_type == type::exact) {
+      _type = 1;
+    } else if (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_RPipeTy<_dataT> _RPipe =
+        __spirv_CreatePipeFromPipeStorage_read<_dataT>(&m_Storage);
+    _dataT TempData;
+    __latency_control_bl_read_wrapper(_RPipe, &TempData, _anchor_id,
+                                      _target_anchor, _type, _cycle);
+    return TempData;
+#else
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  // Writing to pipe is lowered to SPIR-V instruction OpWritePipe via SPIR-V
+  // friendly LLVM IR.
+  template <class... _Params> static void write(const _dataT &_Data) {
+#ifdef __SYCL_DEVICE_ONLY__
+    static constexpr auto _anchor_id =
+        _GetValue<int32_t, latency_anchor_id, _Params...>::value;
+    static constexpr auto _constraint =
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if (_control_type == type::none) {
+      _type = 0;
+    } else if (_control_type == type::exact) {
+      _type = 1;
+    } else if (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_WPipeTy<_dataT> _WPipe =
+        __spirv_CreatePipeFromPipeStorage_write<_dataT>(&m_Storage);
+    __latency_control_bl_write_wrapper(_WPipe, &_Data, _anchor_id,
+                                       _target_anchor, _type, _cycle);
+#else
+    (void)_Data;
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+private:
+  static constexpr int32_t m_Size = sizeof(_dataT);
+  static constexpr int32_t m_Alignment = alignof(_dataT);
+  static constexpr int32_t m_Capacity = _min_capacity;
+#ifdef __SYCL_DEVICE_ONLY__
+  static constexpr struct ConstantPipeStorage m_Storage = {m_Size, m_Alignment,
+                                                           m_Capacity};
+
+  template <typename _T>
+  static int32_t
+  __latency_control_nb_read_wrapper(__ocl_RPipeTy<_T> Pipe, _T *Data,
+                                    int32_t AnchorID, int32_t TargetAnchor,
+                                    int32_t Type, int32_t Cycle) {
+    return __spirv_ReadPipe(Pipe, Data, m_Size, m_Alignment);
```

Comment:
Should it be new ReadPipe instruction coming from SPIR-V extension?

Reply:
Which new ReadPipe instruction you refer to?
This __spirv_ReadPipe builtin function is used in the original [pipes.hpp header](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/pipes.hpp#L31). This experimental header shouldn't make difference on that.

Reply:
I was expecting to have additional SPIR-V interfaces to implement this extension. Otherwise, how we preserve `int32_t AnchorID, int32_t TargetAnchor, int32_t Type, int32_t Cycle` in SPIR-V?

Reply:
They are preserved as wrapper function's arguments.

Reply:
BTW do we even need read/write pipe calls then?

Reply:
Ah, I see, like this we don't have emulation vs hardware problem during the tests.